src/twisted/test/test_defer.py

```diff
@@ -12,28 +12,55 @@
 import traceback
 import re
 import types
-
-from asyncio import new_event_loop, Future, CancelledError
-
-from twisted.python.reflect import requireModule
-from twisted.python import failure, log
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    Generator,
+    List,
+    Mapping,
+    Optional,
+    Tuple,
+    Type,
+    TYPE_CHECKING,
+    Union,
+    cast,
+)
+
+from asyncio import new_event_loop, AbstractEventLoop, Future, CancelledError
+
+from twisted.python import log
+from twisted.python.failure import Failure
 from twisted.trial import unittest
 from twisted.internet import defer, reactor
+from twisted.internet.defer import (
+    _DeferredListResultListT,
+    _DeferredListSingleResultT,
+    ensureDeferred,
+    Deferred,
+    DeferredFilesystemLock,
+    DeferredList,
+    DeferredLock,
+    DeferredSemaphore,
+    DeferredQueue,
+)
 from twisted.internet.task import Clock
 
 
-contextvars = requireModule("contextvars")
-if contextvars:
-    contextvarsSkip = None
+if TYPE_CHECKING:
+    import contextvars
 else:
-    contextvarsSkip = "contextvars is not available"
+    try:
+        import contextvars
+    except ImportError:
+        contextvars = None
 
 
-def ensuringDeferred(f):
+def ensuringDeferred(f: Callable[..., Any]) -> Callable[..., Any]:
     @functools.wraps(f)
-    def wrapper(*args, **kwargs):
+    def wrapper(*args: object, **kwargs: object) -> Deferred[Any]:
         result = f(*args, **kwargs)
-        return defer.ensureDeferred(result)
+        return ensureDeferred(result)
 
     return wrapper
```

Comment:
This decorator is only applied to coroutines, so it could use `Deferred.fromCoroutine()` instead of `ensureDeferred()`. That simplifies its signature compared to `ensureDeferred()` which can take 3 different types:

```py
def ensuringDeferred(
    f: Callable[..., Coroutine[Deferred[_T], Any, _T]]
) -> Callable[..., Deferred[_T]]:
    @functools.wraps(f)
    def wrapper(*args: object, **kwargs: object) -> Deferred[_T]:
        coro = f(*args, **kwargs)
        return Deferred.fromCoroutine(coro)

    return wrapper
```

It's only used 3 times though and all those coroutines have `None` as the result type, so it's probably not all that important to get precise typing here, unless it could be shared with other test modules later.

Reply:
Done, thanks!